lookout/style/format/features.py

```diff
@@ -248,9 +435,28 @@ def extract_features(self, files: Iterable[File], lines: List[List[int]]=None
         offset = 0
         for (vnodes, parents, file_lines), partial_labels in zip(parsed_files, labels):
             offset = self._inplace_write_vnode_features(vnodes, parents, file_lines, offset, X, vn)
+        if self.roles_preprocess != "identity":
+            X = self.encode_roles(X)
+        self._log.info("Features number: %d" % X.shape[1])
         return X, y, vn
 
-    def _classify_vnodes(self, nodes: Iterable[VirtualNode], path: str) -> Iterable[VirtualNode]:
+    def encode_roles(self, X: numpy.ndarray) -> coo_matrix:
+        """
+        Encodes all roles features in data matrix X to its one hot encoding representation.
+        Fits the model if it is unfitted.
+        :param X: Data matrix.
+        :return: New sparse data matrix.
+        """
+        # It is not possible to have negative values to run OneHotEncoder, so here is dirty hack
+        X += 1
```

Comment:
This dirty hack works since we use trees and they are indifferent to strictly monotonic transformations of their features but it would still be good to actually only +1 the categorical columns so that we can interpret stuff later on. It would even be better to not touch anything and just modify tokens.py and roles.py to start at 1 instead of 0.

Reply:
> since we use trees and they are indifferent to strictly monotonic transformations

Yup, that why I use it. I like the last suggestion! Will implement